typedargs/annotate.py

```diff
@@ -264,6 +266,8 @@ def annotated(func, name=None):
     """
 
     if hasattr(func, 'metadata'):
+        if name is not None:
+            func.metadata = AnnotatedMetadata(func, name)
```

Comment:
I'm not 100% sure that it matters but I would recommend just overriding the name rather than reloading the whole metadata object.  It should be doable as just a `func.metadata.name = name` update.  See:
https://github.com/iotile/typedargs/blob/master/typedargs/metadata.py#L73

I think that with this change the following situation would break:

1. `@context` is called on base class to set its name
2. `@param` is also called on base class to mark parameter types of `__init__`
3. subclass inherit from base class without touching the `__init__` function .
4. `@context` is called on the subclass to adjust its name.

I think the call to `@context` in 4 will blow away the parameter types declared in 2, whereas we actually just want to change the name.

Reply:
Whjen I went to doing name = name, it actually overwrote the base class's name too. I think that creating a new AnnotatedMetadata adds a new entry, while overwriting name just masks the one we already created instead of creating a new one:
```
(ota_dev) Matts-Macbook-Pro:arch_gateway mrunchey$ iotile hw --port=virtual:tile_based@./tasks.json
(HardwareManager) connect 1
(HardwareManager) get 10
(MQTTForwarderProxy) back
(HardwareManager) get 11
(MQTTForwarderProxy) back
(HardwareManager) 
```

Reply:
It's also possible I'm defining an unrealistic set of tasks. That is, there shouldn't be nested contexts alive at the same time (`Task` shouldn't actually be a real proxy to be created, and we should always subclass it and context there). For context, I'm creating the above behavior with the following tile definition:
```
            {
                "address": 10,
                "name": "service_delegate",
                "args":
                {
                    "name": "task",
                    "service": "task_proxy"
                }
            },
            {
                "address": 11,
                "name": "service_delegate",
                "args":
                {
                    "name": "mqttfw",
                    "service": "mqtt_forwarder"
                }
            }
```

Reply:
However, if I use multiple objects that both inherit from the base `TaskProxy` class, they'll get the name of the last one that's defined:
```
(ota_dev) Matts-Macbook-Pro:arch_gateway mrunchey$ iotile hw --port=virtual:tile_based@./tasks.json
(HardwareManager) connect 1
(HardwareManager) get 10
(MQTTForwarderProxy) back
(HardwareManager) get 11
(MQTTForwarderProxy) 
```

If I remove the `@context` from `TaskProxy`, there are no longer nested context calls, but that renders the functional behavior of this PR moot.

Reply:
No, you're right.  I think we may actually need some test coverage to make sure that the `param` decorators from the `__init__` method aren't getting obliterated:

```python
@param('arg1', 'bool')
@context('Base')
class Base:
   def __init__(self, arg1):
        assert arg1 is bool

@context('Sub')
class Sub(Base):
     pass

assert Sub('false') is False
```

Reply:
Actually, possibly not.  The example I was thinking of is this one:
https://github.com/iotile/coretools/blob/master/iotilecore/iotile/core/hw/hwmanager.py#L42

But it does not nest `@context` and `@param`:
https://github.com/iotile/coretools/blob/master/iotilecore/iotile/core/hw/hwmanager.py#L81

Reply:
I think your approach should be good.